src/storage/src/source/persist_source.rs

```diff
@@ -348,11 +348,24 @@ where
             while let Some((cap, data)) = input.next() {
                 let cap = cap.retain();
                 for (_idx, batch) in data.iter() {
-                    let mut updates = read
+                    let parts = read
                         .fetch_batch(batch.clone())
                         .await
                         .expect("shard_id generated for sources must match across all workers");
 
+                    // WIP keep this alloc around?
+                    let mut updates = Vec::new();
+                    for part in parts {
+                        for update in part {
+                            // Allow OperatorBuilderExt an opportunity to yield
```

Comment:
@aljoscha any idea if this is actually how to do this?

Reply:
I think maybe not, because there is no `true` await point in the loop. But I don't know for sure. Only thing I could do is sprinkle some log statements and observe. :sweat_smile: 

The way a "classic" timely operator would solve this is store the work that it needs to do in some stash. When it's activated it will work on that for a bit, activate itself using an `Activator`, and return. Might be that we have to do a similar thing, without using the `async_op!` macro.

Reply:
good(?) news!

I read the build_async operator code very closely and it really seems like it yields to timely after every non-ready call to `poll`. The future being polled is just our async block wrapped in a Fuse (which I manually checked for any sort of multi-poll-per-poll nonsense), so the fix here should be as easy as putting any await call that isn't immediately ready when we want to pause. Then we don't have to do anything fancy with keeping track of where we are, that's already baked into the Future state machine compiled for us by async/await.

I also tested this in practice and indeed it works. Given the following bit (plus some bonus logging scattered around):

```
                        update_session.give(update);
                        tracing::info!("WIP {} gave an update to the session", fetcher_name);
                        let mut polled = false;
                        let () = futures::future::poll_fn(move |cx| match polled {
                            true => Poll::Ready(()),
                            false => {
                                polled = true;
                                cx.waker().wake_by_ref();
                                Poll::Pending
                            }
                        })
                        .await;
```

We get the following logs:

```
scheduled operator "persist_source 0: part fetcher u1"
persist_source 0: part fetcher u1 finished fetching part LeasedBatchPart { shard_id: ShardId(a39d3cf4-abd6-4494-b770-6fe851ff964d), reader_id: ReaderId(13381028-8351-4086-8638-c4565d752b1a), metadata: Snapshot { as_of: [[133, 198, 170, 240, 130, 1, 0, 0]] }, desc: Description { lower: Antichain { elements: [1661895099905] }, upper: Antichain { elements: [1661895100038] }, since: Antichain { elements: [0] } }, key: PartialBatchKey("w0205d51b-f032-41ca-804b-ba10e7889f2a/ped162a62-c393-4068-b517-36938b6513fd"), leased_seqno: Some(SeqNo(7)) }
persist_source 0: part fetcher u1 gave an update to the session
async returned from poll true "persist_source 0: part fetcher u1"
returning from operator "persist_source 0: part fetcher u1"
scheduled operator "persist_source 0: part fetcher u1"
persist_source 0: part fetcher u1 gave an update to the session
async returned from poll true "persist_source 0: part fetcher u1"
returning from operator "persist_source 0: part fetcher u1"
scheduled operator "persist_source 0: part fetcher u1"
persist_source 0: part fetcher u1 gave an update to the session
async returned from poll true "persist_source 0: part fetcher u1"
returning from operator "persist_source 0: part fetcher u1"
scheduled operator "persist_source 0: part fetcher u1"
```

Reply:
Sounds good! You didn't yet update this PR to do that, right? Will you wait until you finish the "split `LeasedBatch` into parts" work?

Reply:
Correct, I haven't. I was going to wait until I can rebase this on top of #14525, yes, but also I was going to poke frank and/or mortiz for gut instincts on what heuristic we should use for when we yield :)